### ISSUE TITLE
PHP 7.2 incompatibly in TControl

### ISSUE THREAD

**Issue by jcass77** [pradosoft/prado#706]
PHP 7.2 introduces a backward-incompatible change that issues a warning when [counting non-countable types](https://www.php.net/manual/en/migration72.incompatible.php).

This breaks `TControl` - a similar issue was reported some time ago as part of #663.

```php
[Warning] count(): Parameter must be an array or an object that implements Countable (@line 1702 in file /var/www/html/atrade/test/sharekiosk-portal/protected/vendor/pradosoft/prado/framework/Web/UI/TControl.php).

Source File

${PradoFramework}/Web/UI/TControl.php (1702)

1691:     {
1692:         if (isset($this->_rf[self::RF_ADAPTER])) {
1693:             $this->_rf[self::RF_ADAPTER]->saveState();
1694:         } else {
1695:             $this->saveState();
1696:         }
1697:         $needViewState = ($needViewState && !($this->_flags & self::IS_DISABLE_VIEWSTATE));
1698:         $state = [];
1699:         if ($this->getHasControls()) {
1700:             foreach ($this->_rf[self::RF_CONTROLS] as $control) {
1701:                 if ($control instanceof TControl) {
1702:                     if (count($tmp = &$control->saveStateRecursive($needViewState))) {
1703:                         $state[$control->_id] = $tmp;
1704:                     }
1705:                 }
1706:             }
1707:         }
1708:         if ($needViewState && !empty($this->_viewState)) {
1709:             $state[0] = &$this->_viewState;
1710:         }
1711:         if (isset($this->_rf[self::RF_CONTROLSTATE])) {
1712:             $state[1] = &$this->_rf[self::RF_CONTROLSTATE];
1713:         }
1714:         return $state;
```

**Comment by ctrlaltca** [comment]
Hi and thank you for the bug report.
`TControl::saveStateRecursive()` is supposed to always return an array.
What value do you get instead, `null`?
This could be caused by a custom control reimplementing `saveStateRecursive()` or by some bug in`TOutputCache`, if used.

**Comment by jcass77** [comment]
I am using `TOutputCache`, and I think that it [returns some serialised output](https://github.com/pradosoft/prado/blob/9d2c3dbaf7f4e71e774b5b0411c22a078798e517/framework/Web/UI/WebControls/TOutputCache.php#L239-L242) that is not Countable.

Could that be the issue?

**Comment by ctrlaltca** [comment]
That's definitely what's causing the issue. I'll try to fix it asap.

**Comment by ctrlaltca** [comment]
After a few tests about the best way to handle this, I've not found an example where the serialization is needed for the page to work. So I've just removed it, see https://github.com/pradosoft/prado/pull/707
Can you please give it a test run and see if it works for you, too?

**Comment by jcass77** [comment]
I did a quick test and the change in #707 seems ok.

**Comment by ctrlaltca** [comment]
Thank you for the confirmation; anyway I did not include the fix in the recently released 4.1.1 to ensure proper testing of #707.
